File: Civitas-Ruby/lib/controlador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'vista_textual'
require_relative 'civitas_juego'
require_relative 'operaciones_juego'
require_relative 'operacion_inmobiliaria'
require_relative 'gestiones_inmobiliarias'

module Civitas
  class Controlador
    def initialize(juego, vista)
      @juego = juego
      @vista = vista
    end
    
    def juega     
      @vista.set_civitas_juego(@juego)
      
      while !@juego.finaldeljuego
        @vista.actualizar_vista
        @vista.pausa
        siguiente = @juego.siguientepaso
        @vista.mostrar_siguiente_operacion(siguiente)
        
        if siguiente != Civitas::Operaciones_juego::PASAR_TURNO
          @vista.mostrar_eventos
        end
        
        if !@juego.finaldeljuego
          case siguiente
          when Civitas::Operaciones_juego::COMPRAR
            r = @vista.comprar
            if r == Civitas::Respuestas::SI
              @juego.comprar
              @juego.siguientepasocompletado(siguiente)
            end
          when Civitas::Operaciones_juego::GESTIONAR
            @vista.gestionar
            lista_gestion = [Civitas::GestionesInmobiliarias::CANCELAR_HIPOTECA,
              Civitas::GestionesInmobiliarias::CONSTRUIR_CASA,
              Civitas::GestionesInmobiliarias::CONSTRUIR_HOTEL,
              Civitas::GestionesInmobiliarias::HIPOTECAR,
              Civitas::GestionesInmobiliarias::TERMINAR,
              Civitas::GestionesInmobiliarias::VENDER
            ]
            
            operacion = OperacionInmobiliaria.new(lista_gestion[@vista.iGestion], @vista.iPropiedad)
            
            case opeacion.iGestion
            when Civitas::GestionesInmobiliarias::TERMINAR
              @juego.siguientepasocompletado(siguiente)
            when Civitas::GestionesInmobiliarias::CANCELAR_HIPOTECA
              @juego.cancelarhipoteca(operacion.num_propiedad)
            when Civitas::GestionesInmobiliarias::CONSTRUIR_CASA
              @juego.construircasa(operacion.num_propiedad)
            when Civitas::GestionesInmobiliarias::CONSTRUIR_HOTEL
              @juego.construirhotel(operacion.num_propiedad)
            when Civitas::GestionesInmobiliarias::HIPOTECAR
              @juego.hipotecar(operacion.num_propiedad)
            when Civitas::GestionesInmobiliarias::VENDER
              @juego.vender(operacion.num_propiedad)
            end
            
          when Civitas::Operaciones_juego::SALIR_CARCEL
            if @vista.salir_carcel == Civitas::SalidasCarcel::TIRANDO
              @juego.salircarceltirando
            elsif @vista.salir_carcel == Civitas::SalidasCarcel::PAGANDO
              @juego.salircarcelpagando
            end
            
            @juego.siguientepasocompletado(siguiente)
          end
        else
          #He cambiado la visibilidad del ranking de jugadores
          @juego.ranking
        end
      end
    end
  end
end

File: Civitas-Java/Civitas/src/main/java/GUI/Controlador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package GUI;

import civitas.CivitasJuego;
import civitas.GestionesInmobiliarias;
import civitas.OperacionInmobiliaria;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;

/**
 * @author elena
 * @date 25/10/2019
 */

public class Controlador {
    private CivitasJuego juego;
    private CivitasView vista;
    
    Controlador(CivitasJuego juego, CivitasView vista){
        this.juego = juego;
        this.vista = vista;
    }
    
    void juega(){
        vista.setCivitasJuego(juego);
        
        while(!juego.finalDelJuego()){
            vista.actualizarVista();
            //vista.pausa();           
            OperacionesJuego siguiente = juego.siguientePaso();
            
            vista.mostrarSiguienteOperacion(siguiente);
            if(siguiente != OperacionesJuego.PASAR_TURNO){
                vista.mostrarEventos();
            }
            
            if(!juego.finalDelJuego()){
                switch(siguiente){
                    case COMPRAR:
                        Respuestas r = vista.comprar();
                        
                        if(r == Respuestas.SI){
                            juego.comprar();                                                          
                        }
                        
                        juego.siguientePasoCompletado(siguiente);  
                        
                    break;
                    case GESTIONAR:
                        vista.gestionar();
                        OperacionInmobiliaria operacion_inmo = new OperacionInmobiliaria(GestionesInmobiliarias.values()[vista.getGestion()], vista.getPropiedad());     
                        
                        switch(operacion_inmo.getGestion()){
                            case TERMINAR:
                                juego.siguientePasoCompletado(siguiente);
                            break;
                            case CANCELAR_HIPOTECA:
                                juego.cancelarHipoteca(operacion_inmo.getNumPropiedad());
                            break;
                            case CONSTRUIR_CASA:
                                juego.construirCasa(operacion_inmo.getNumPropiedad());
                            break;
                            case CONSTRUIR_HOTEL:
                                juego.construirHotel(operacion_inmo.getNumPropiedad());
                            break;
                            case HIPOTECAR:
                                juego.hipotecar(operacion_inmo.getNumPropiedad());
                            break;
                            case VENDER:
                                juego.vender(operacion_inmo.getNumPropiedad());
                            break;
                        }
                    break;
                    case SALIR_CARCEL:                        
                        if(vista.salirCarcel() == SalidasCarcel.PAGANDO)
                            juego.salirCarcelPagando();                        
                        else if(vista.salirCarcel() == SalidasCarcel.TIRANDO)
                            juego.salirCarcelTirando();
                        
                        juego.siguientePasoCompletado(siguiente);
                    break;
                }
            }
            else{
                //He cambiado la visibilidad del ranking de jugadores
                juego.ranking();
            }
        }
    }
}
